### PR TITLE
Update developer guide for model, find feature and budget feature

### PR DESCRIPTION
**Pull request by zhaojj2209**
Partially addresses #297.

Updated developer guide Model section, find feature and budget feature.

### CODE REVIEW

**Review by wltan — CHANGES_REQUESTED** [review]
image
Should this be an interface instead?

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -128,15 +130,22 @@ This abstraction is maintained so that the implementation of other components (`
...
-The `ModelManager` has three `FilteredList` fields which point to the same `ObservableList` obtained from `FinanceTracker::getTransactionList`.
-The `Predicate` fields in the three `FilteredList` fields are set such that:
-* `filteredTransactions` shows a view of all `Transaction` objects
-* `filteredExpenses` shows a view of all `Transaction` objects of type `Expense`
-* `filteredIncomes` shows a view of all `Transaction` objects of type `Income`
+The `ModelManager` has the following fields:
+* three `FilteredList` fields which point to the same `ObservableList` obtained from `FinanceTracker::getTransactionList`. The `Predicate` fields in the three `FilteredList` fields are set such that:
+    * `filteredTransactions` shows a view of all `Transaction` objects
+    * `filteredExpenses` shows a view of all `Transaction` objects of type `Expense`
+    * `filteredIncomes` shows a view of all `Transaction` objects of type `Income`
```
Comment (ianyong) [inline docs/DeveloperGuide.md:143]: Add full stops to be consistent with lines 134 to 137?

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -128,15 +130,22 @@ This abstraction is maintained so that the implementation of other components (`
...
 
-The motivation behind having three lists is due to the fact that there are three tabs in the user interface, each having its own list while at the same time retrieving data from the same transaction list.
+    The motivation behind having three lists is due to the fact that there are three tabs in the user interface, each having its own list while at the same time retrieving data from the same transaction list.
```
Comment (ianyong) [inline docs/DeveloperGuide.md:145]:
```suggestion
    The motivation behind having three lists is due to the fact that there are three tabs in the user interface which display transactions, each having its own list while at the same time retrieving data from the same transaction list.
```

Sounds like we have only three tabs in the user interface.

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -167,28 +176,63 @@ Classes used by multiple components are in the `ay2021s1_cs2103_w16_3.finesse.co
 
 This section describes some noteworthy details on how certain features are implemented.
 
-### Find transactions
+### Transactions
+
+#### Find transactions feature
+##### Overview
```
Comment (ianyong) [inline docs/DeveloperGuide.md:182]: image

Are the consecutive headings intended?

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -167,28 +176,63 @@ Classes used by multiple components are in the `ay2021s1_cs2103_w16_3.finesse.co
...
 
-* Having separate command parsers for each tab in which the find command can be input, e.g. `FindTransactionCommandParser`, `FindExpenseCommandParser` and `FindIncomeCommandParser`.
+The following activity diagram summarizes what happens when the user executes a new find command:
 
-### Programmatically switch selected tab
+![Activity diagram for executing the find command](images/FindActivityDiagram.png)
+
+##### Design considerations
+
+The alternative implementations considered, as well as the rationale behind our current implementation is as follows:
```
Comment (ianyong) [inline docs/DeveloperGuide.md:228]:
```suggestion
The alternative implementations considered, as well as the rationale behind our current implementation are as follows:
```

Not too sure about this.

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -208,19 +252,60 @@ Alternatives considered:
...
 
-The monthly budgeting feature is implemented via `SetSavingsGoalCommand` as well as `MonthlySavingsGoal`.
+##### Implementation of feature
 
-[Coming soon]
+The find transactions feature is implemented via `MonthlyBudget`, which contains the following fields:
+
+* Two `ObjectProperty<Amount>` fields for `monthlyExpenseLimit` and `monthlySavingsGoal`.
+* Two `ObjectProperty<CalculatedAmount>` fields for `remainingBudget` and `currentSavings`.
+* Three c fields for `monthlyExpenses`, `monthlyIncomes` and `monthlySavings`, which are used for the [Analytics feature](#analytics).
```
Comment (ianyong) [inline docs/DeveloperGuide.md:271]: `Three c fields`?

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -208,19 +252,60 @@ Alternatives considered:
...
+* Sums the amounts of all `Expense`s in the `TransactionList` dated in the current month, subtracts it from the sum of all `Income`s dated in the current month, and sets it as the `currentSavings`.
+* For the past number of months indicated by `numOfMonths` (including the current month), calculate the total expenses, total incomes and total savings for each month and add them to the corresponding `ObservableList<CalculatedAmount>`.
+
+The commands that trigger a call of `calculateBudgetInfo` are as follows:
+* `AddExpenseCommand`, `AddIncomeCommand`, `DeleteExpenseCommand` and `DeleteIncomeCommand`, which change the total expenses/incomes of the `Transaction`s in the `TransactionList`.
+* `EditExpenseCommand` and `EditIncomeCommand`, if the `Amount` is edited.
+* `SetExpenseLimitCommand`, which changes the `Amount` in `monthlyExpenseLimit`.
+* `SetSavingsGoalCommand`, which changes the `Amount` in `monthlySavingsGoal`.
+`calculateBudgetInfo` is also called by `Storage` whenever the finance tracker data is loaded upon startup.
```
Comment (ianyong) [inline docs/DeveloperGuide.md:287]: Should this be a bullet point or otherwise be on a new line?

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -208,19 +252,60 @@ Alternatives considered:
...
-### [Proposed] Frequent Expense Feature
+![Sequence diagram for executing the `setel a/500` command](images/SetExpenseLimitSequenceDiagram.png)
+
+The following activity diagram summarizes what happens when the user executes any command that changes the calculated values in `MonthlyBudget` (e.g. setting expense limit):
+
+![Activity diagram for executing any command that changes the budget](images/BudgetActivityDiagram.png)
+
+##### Design considerations
+
+The alternative implementations considered, as well as the rationale behind our current implementation is as follows:
```
Comment (ianyong) [inline docs/DeveloperGuide.md:302]: Similar to above.

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -369,40 +457,294 @@ testers are expected to do more *exploratory* testing.
...
+1. Switching tabs
+
+    1. Test case: `tab 2`
+       Expected: UI switches to the Expenses tab.
+
+    1. Test case: `tab overview`
+       Expected: UI does not switch tabs. An error message is shown as the command format is invalid.
+
+    1. Test case: `tab 5`
+       Expected: UI does not switch tabs. An error message is shown as the specified tab does not exist.
```
Comment (ianyong) [inline docs/DeveloperGuide.md:491]: Missing `<br>`s here.

**Review by ianyong — CHANGES_REQUESTED** [review]

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (zhaojj2209) [reply docs/DeveloperGuide.md:182]: The headings are intended as the Transactions section is split into subsections by feature, while for other sections (e.g. Budgeting), the whole section is one huge feature and need not be split into subsections.

**Review by wltan — APPROVED** [review]
LGTM!

**Review by wltan — APPROVED** [review]
LGTM!

**Review by ianyong — APPROVED** [review]
LGTM!

**Pull request merged**